packages/integrations/mdx/src/astro-data-utils.ts

```diff
@@ -0,0 +1,78 @@
+import { name as isValidIdentifierName } from 'estree-util-is-identifier-name';
+import type { VFile } from 'vfile';
+import type { MdxjsEsm } from 'mdast-util-mdx';
+import type { MarkdownAstroData } from 'astro';
+import type { Data } from 'vfile';
+import { jsToTreeNode } from './utils.js';
+
+export function remarkInitializeAstroData() {
+	return function (tree: any, vfile: VFile) {
+		if (!vfile.data.astro) {
+			vfile.data.astro = { frontmatter: {} };
+		}
+	};
+}
+
+export function rehypeApplyFrontmatterExport(pageFrontmatter: object, exportName = 'frontmatter') {
+	return function (tree: any, vfile: VFile) {
+		if (!isValidIdentifierName(exportName)) {
+			throw new Error(
+				`[MDX] ${JSON.stringify(
+					exportName
+				)} is not a valid frontmatter export name! Make sure "frontmatterOptions.name" could be used as a JS export (i.e. "export const frontmatterName = ...")`
+			);
+		}
+		const frontmatter = { ...pageFrontmatter, ...safelyGetAstroData(vfile.data).frontmatter };
```

Comment:
Can we be more defensive here? 

It would be better if `pageFrontmatter` came last to override the injected frontmatter.
As a user I would probably also want a warning when conflicts are detected, because a plugin tried to override some explicit frontmatter value that I wrote.

Reply:
I like that! Let's do that instead :thu